tasks/grunt-karma.js

```diff
@@ -71,22 +71,27 @@ module.exports = function (grunt) {
       data.configFile = path.resolve(data.configFile)
     }
 
-    if (data.files || options.files) {
-      data.files = [].concat.apply(options.files || [], this.files.map(function (file) {
-        return file.src.map(function (src) {
-          var obj = {
-            pattern: src
-          }
-          var opts = ['watched', 'served', 'included']
-          opts.forEach(function (opt) {
-            if (opt in file) {
-              obj[opt] = file[opt]
+    if (this.files.length || options.files) {
+      data.files = [].concat.apply(
+        // For the 'files' option, we support nested arrays as a convenient
+        // way to specify files without needing the user to concat or
+        // flatten them ahead of time.
+        _.flattenDeep(options.files || []),
```

Comment:
I think this would be clearer and simpler as 
```js
// three lines of comment
data.files = _.flattenDeep(options.files || []).concat(
  this.files.map(() => {
    return file.src.map((src) => {
    ...
```

Reply:
I'd like that, and it avoids an extra array and func-apply indirection. However, it doesn't work as-is because while concat() supports an array as parameter, it only unwraps one layer, and we need exactly two layers. So we need the `concat.apply()` to ensure each array is its own parameter.

Alternatively, while we don't need arbitrary depth on the mapped files, we could go back to passing the entire expression to `flattenDeep`, by adding two parentheses:

```js
 // three lines of comment
data.files = _.flattenDeep((options.files || []).concat(
  this.files.map(() => {
    return file.src.map((src) => {
    ...
))
```

Thoughts?

Reply:
With the proposed snippet, the tests fail as follows:

```
 Invalid pattern [object Object]!
	Object is missing "pattern" property.
```

This is because `data.files` will contain nested arrays from the map() inside the map().

> ```js
> console.log([].concat.apply([1,2], [3,4]));
> console.log([1,2].concat([3,4]));
> ```
> aren't these both [1,2,3,4]?

Yes, their outcome happens to be identical, but their logic is not identical. Instead, these two are identical:

```js
[].concat.apply([1,2], [3,4]);
[1,2].concat(3, 4);
```

Notice that concat() is called with numbers, not arrays. Function#bind takes two parameters, `thisValue` and `argumentsList`.

The following are also identical, and matches grunt-karma code:

```js
[].concat.apply([1, 2], [[3],[4]]);
[1,2].concat([3], [4]);
//> [ 1, 2, 3, 4 ]
```

But, the proposed snippet would do:
```
[1,2].concat([[3], [4]]);
//> [ 1, 2, [ 3 ], [ 4 ]
```